src/type.js

```diff
@@ -240,8 +243,16 @@ function fireInputEventIfNeeded({
   newValue,
   newSelectionStart,
   eventOverrides,
+  textToBeTyped,
 }) {
   const prevValue = currentElement().value
+
+  if (
+    currentElement().type === 'date' &&
+    !isValidDateValue(currentElement(), textToBeTyped)
+  )
+    return {prevValue}
```

Comment:
How about keeping this special case outside of `fireInputEventIfNeeded` and instead add the following at the end of the `calculateNewValue` util?

```js
if (element.type === 'date' && !isValidDateValue(element, newValue)) {
  newValue = value
}
```

This would keep things a bit easier to reason about.

Reply:
This would not avoid a `change` event to be triggered. We could move the condition to outside, in `typeCharacter` scope, but the linter will throw a `complexity` violation.

>As for the events, just to make it clear: date input just triggers an input event (plus a change event) on a valid date according to the user-event playground

Reply:
It should actually do the same as the current implementation. We already check for `newValue = currentElement.value()` in `fireInputEventIfNeeded`. So when `calculateNewValue` returns the elements current value, the input event will not be fired.

Regarding the `change` event: right now it looks like we only fire it inside `typeCharacter` if we have a valid date on a date input. This behavior would not be affected.

Reply:
My bad, I typo the events, I meant the `input` event: it should be avoided when a date is not valid.

Oh, I overlooked that, I think it might work If we change the `calculateNewValue` function to receive the new value as you have said, though I'm now sure about this move since `calculateNewValue` is more concerning about the new value in a "selection aware". What do you think about it, @kentcdodds?

Thanks

Reply:
I'm ok with the suggested change

Reply:
Done. I've to agree, it's better. :)